k8srm-prototype/pod-spec-examples.yaml

```diff
@@ -0,0 +1,127 @@
+# EXAMPLE SNIPPETS FROM PodSpec
+#
+# NOTE: We want to allow a set of "must satisfy all"
+# and multiple sets of "must satisfy one of", and
+# we want to allow matchAttributes/distinctAttributes
+# across all of the selected devices.
+#
+# As written here, each entry in deviceClaims is claimName that
+# points to an existing claim or a template with:
+# - ObjectMeta
+# - MatchAttributes (and maybe DistinctAttributes?) that apply across all
+#   claims in the Claims list
+# - Claims list, each of which is either a claim (inlined), or
+#   a "oneOf" list of claims.
+
+# This means we have an extra "  claims:" line in the simplest of cases.  We
+# could eliminate that with equivalent expressiveness, by instead listing
+# "allOf" and "oneOf", where "oneOf" is actually a list-of-lists of claims, but
+# not sure that's better.
+
+
+# Allocates four SR-IOV VFs from any Intel NIC These could even each come from
+# a different NIC (on the same node of course). NICs are not tied to specific
+# containers, so there is no need to name the claim.
+deviceClaims:
+- template:
+    claims:
+    - deviceClass: "intel-sriov-vf"
+      requests:
+        count: 4
+
+# Allocates four SR-IOV VFs from any vendor NIC, making sure all the VFs are on
+# the same PF this requires vendor & PF together to fully qualify a physical
+# card on a node.
+deviceClaims:
+- template:
+    claims:
+    - deviceType: "sriov-vf"
+      count: 4
+      matchAttributes:
+      - "vendor"
+      - "pf"
+
+# Allocates two A100 GPUs on the same PCI root complex. The name is used to
+# attach the associated claims to a container.
+deviceClaims:
+- template:
+    claims:
+    - name: gpus
+      deviceClass: "nvidia-a100"
+      count: 2
+      matchAttributes:
+      - pcie-root
+
+# Allocates two A100 GPUs on the same PCI root complex, but keeps them
+# separated in the API so they can be attached to two differenti containers.
+deviceClaims:
+- template:
+    matchAttributes:
+    - pcie-root
+    claims:
+    - name: gpu-one
+      deviceClass: "nvidia-a100"
+    - name: gpu-two
+      deviceClass: "nvidia-a100"
+
+# Allocates an SR-IOV PF and a an A100 from the same pcie-root. Name is
+# populated here for the PF just to show that it can be, but the expectation is
+# generally for network devices it is not needed. Note ANY vendor PF is fine
+# here; we use deviceType, not deviceClass.
+deviceClaims:
+- template:
+    matchAttributes:
+    - pcie-root
+    claims:
+    - name: gpu
+      deviceClass: "nvidia-a100"
+    - name: nic
+      deviceType: "sriov-pf"
+
+# Allocates either an A100, H100, or two L4s that are on the same pcie root.
+deviceClaims:
+- template:
+    claims:
+    - oneOf:
+      - deviceClass: "nvidia-a100"
+      - deviceClass: "nvidia-h100"
+      - deviceClass: "nvidia-l4"
```

Comment:
In a real setup I do not expect admins to create different classes for these devices. There will be one `nvidia.com.gpu` class and the 'selector' API will be used to select a particular type of GPU (if they care).

Reply:
Ok, as I stated in another comment, we should vet these out in the WG. I hear the opposite from others - do everything in the class, claims will be too hard for most users.

For the "nvidia CRD-based API", I think rendering the output of that as classes rather than ResourceClaimParameters allows us to keep the number of objects in the K8s model lower, with no need for an equivalent of ResourceClaimParameters - just a set of references to config objects, but equally expressive.

Reply:
My experience is that there are three type of users:
1. Give me all of the GPUs on the machine
2. Give me exactly one GPU that has at least X amount of memory (I don't care if its a MIG device or a full GPU)
3. Give me a specific number of GPUs matching some complex set of constraints (GPU type, amount of memory, shared with MPS vs time-slicing, etc.)

Meaning that we should make it as easy as possible to express 1 and 2, but leave the door open for power users to express 3. Meaning (for GPUs at least), I'd only expect 3 classes, ever -- one for each of the 3 cases above. Where the generic `nvidia.com.gpu`  mentioned above  maps to the third one -- i.e. the one giving the user full control of asking for exactly what they want.

Reply:
We actually don't really have a way to do "1", unless of course you already know *how many* there are. Is that adequate? I think it probably is.

For "2", I have also heard "give me N GPUs, so long as they together cover this amount of memory", and "give me one of these, or if you don't have it, one of those, or if you don't have it 2 of these".

Reply:
We support 1 in "classic" DRA since the vendor owns the allocation code. I apparently overlooked it when we moved to structured parameters, but it is required in general for being able to deploy "management" containers as daemonsets (different nodes often have a different numbers of GPUs on them and the management containers want to be able to see / monitor all of them).

Reply:
In general, I'm hesitant to assume that the common case will be to have a proliferation of admin-driven resource classes vs. user-driven resource claims. My experience is that users packaging up an application (usually as a helm chart with a bunch of pods / claims predefined) want this level of control.